crypto/src/test_utils/test_conversation/operation_guard.rs

```diff
@@ -0,0 +1,240 @@
+    }
+
+    /// Propoagate the state from the commit to the [TestConversation].
+    /// Needed if you notify members manually and never call [Self::notify_member].
+    pub async fn process_member_changes(mut self) -> Self {
+        // Do the following for each proposal that is still pending and the latest commit:
+        // In case of a remove, an external join or an add operation, update the member list
+        // of the test conversation.
+        for operation in self
+            .conversation
+            .proposals
+            .iter()
+            .chain(std::iter::once(&self.operation))
+        {
+            match operation {
+                TestOperation::Update => {}
+                TestOperation::Remove(member) => {
+                    if !self.conversation().is_member(member).await {
+                        // because we're eagerly pushing proposals into the list of operations to process,
+                        // it's possible that we have duplicate operations.
+                        continue;
+                    }
+                    let member_idx = self.conversation().member_index(member).await;
+                    self.conversation.members.remove(member_idx);
+                }
+                TestOperation::ExternalJoin(joiner) => {
+                    // If this is a rejoin, don't touch the member list
+                    if !self.conversation().is_member(joiner).await {
+                        self.conversation.members.push(joiner);
+                    }
+                }
+                TestOperation::Add(AddGuard {
+                    new_members: invited_members,
+                }) => {
+                    if self.conversation().are_members(invited_members.clone()).await {
+                        continue;
+                    }
+                    let welcome_message = self
+                        .conversation()
+                        .actor()
+                        .mls_transport()
+                        .await
+                        .latest_commit_bundle()
+                        .await
+                        .welcome
+                        .expect("we're processing an add operation, so there must be a welcome message");
+                    // Process welcome message on receiver side
+                    for new_member in invited_members.iter() {
+                        new_member
+                            .transaction
+                            .process_welcome_message(
+                                welcome_message.clone().into(),
+                                self.conversation.case.custom_cfg(),
+                            )
+                            .await
+                            .unwrap();
+                    }
+                    // Then update the member list
+                    self.conversation.members.reserve(invited_members.len());
+                    self.conversation.members.extend(invited_members);
+                }
+            }
```

Comment:
Could you please move this `match` block into a separate function (e.g. `process_member_change`) so it only does one thing? Then this function can just be a simple iteration over all proposal + self.